### PR TITLE
Support token auth for /api/* resources

### LINKED ISSUE #189: Option to support a shared secret token auth scheme
Right now, the tmpnb spawn, stats, etc. end points are wide open for anyone to request access to a new container. This is great for open deployments, not so great for cases where only certain actors should be allowed to request containers. An optional API key scheme would do the trick here. I suggest keeping with what we did in jupyter-incubator/kernel_gateway#3, generate a single shared token on startup, log it,  require it in the form:

```
Authorization: token <the api key>
```

I'll probably wind up tackling this in a PR. Opening it so I don't forget. Did discuss it briefly with @rgbkrk

### PR DESCRIPTION
**Pull request by parente**
* Accept optional API key from API_AUTH_TOKEN env var
* Disable human-facing handlers when API token is provided
* Check API token in BaseHandler
* Update README

Fixes #189

### CODE REVIEW

**Comment by parente** [comment]
Notebook "test case" demonstrating how it works: https://gist.github.com/parente/cef02128c4d581f11d9a

**Comment by rgbkrk** [comment]
YES!

File: orchestrate.py [hunk 1]
```diff
@@ -306,27 +316,33 @@ def main():
...
-        (r"/", LoadingHandler),
-        (r"/spawn/?(/user/\w+(?:/.*)?)?", SpawnHandler),
-        (r"/spawn/((?:notebooks|tree)(?:/.*)?)", SpawnHandler),
         (r"/api/spawn/?", APISpawnHandler),
-        (r"/(user/\w+)(?:/.*)?", LoadingHandler),
-        (r"/((?:notebooks|tree)(?:/.*)?)", LoadingHandler),
-        (r"/api/stats/?", APIStatsHandler),
-        (r"/stats/?", RedirectHandler, {"url": "/api/stats"}),
-        (r"/info/?", InfoHandler),
+        (r"/api/stats/?", APIStatsHandler)
```
Comment (rgbkrk) [inline orchestrate.py]: Ok, so if there's an API Token then we only have the /api/stats and /api/spawn endpoints. Seems alright.

File: orchestrate.py [hunk 2]
```diff
@@ -48,6 +48,12 @@ def prepare(self):
         if self.allow_headers:
             self.set_header("Access-Control-Allow-Headers", self.allow_headers)
 
+        # Confirm the client authorization token if an api token is configured
+        if self.api_token is not None:
+            client_token = self.request.headers.get('Authorization')
+            if client_token != 'token %s' % self.api_token:
+                return self.send_error(401)
```
Comment (minrk) [inline orchestrate.py]: I believe this should be `raise web.HTTPError(403)`

**Pull request merged**